sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala

```diff
@@ -30,8 +36,156 @@ case class GpuJsonToStructs(
+        (0 until col.getNumChildren).foreach { i =>
+          val child = col.getChildColumnView(i)
+          structFields += StructField("", getSparkType(child))
+        }
+        StructType(structFields)
+      case t => throw new IllegalArgumentException(
+        s"GpuJsonToStructs currently cannot process CUDF column of type $t.")
+    }
+  }
+
   override protected def doColumnar(input: GpuColumnVector): cudf.ColumnVector = {
-    MapUtils.extractRawMapFromJsonString(input.getBase)
+    schema match {
+      case _: MapType =>
+        MapUtils.extractRawMapFromJsonString(input.getBase)
+      case struct: StructType => {
+        // We cannot handle all corner cases with this right now. The parser just isn't
+        // good enough, but we will try to handle a few common ones.
+        val numRows = input.getRowCount.toInt
+
+        // Step 1: verify and preprocess the data to clean it up and normalize a few things
+        // Step 2: Concat the data into a single buffer
+        val (isNullOrEmpty, combined) = cleanAndConcat(input.getBase)
+        withResource(isNullOrEmpty) { isNullOrEmpty =>
+          // Step 3: copy the data back to the host so we can parse it.
+          val combinedHost = withResource(combined) { combined =>
+            combined.copyToHost()
+          }
+          // Step 4: Have cudf parse the JSON data
+          val (names, rawTable) = withResource(combinedHost) { combinedHost =>
+            val data = combinedHost.getData
+            val start = combinedHost.getStartListOffset(0)
+            val end = combinedHost.getEndListOffset(0)
+            val length = end - start
+
+            withResource(cudf.Table.readJSON(cudf.JSONOptions.DEFAULT, data, start,
+              length)) { tableWithMeta =>
+              val names = tableWithMeta.getColumnNames
+              (names, tableWithMeta.releaseTable())
+            }
+          }
+
+          // process duplicated field names in input struct schema
+          val fieldNames = processFieldNames(struct.fields.map { field => 
+              (field.name, field.dataType)})
+
+          withResource(rawTable) { rawTable =>
+            // Step 5: verify that the data looks correct
+            if (rawTable.getRowCount != numRows) {
+              throw new IllegalStateException("The input data didn't parse correctly and we read " +
+                  s"a different number of rows than was expected. Expected $numRows, " +
+                  s"but got ${rawTable.getRowCount}")
+            }
+
+            // Step 6: get the data based on input struct schema
+            val columns = fieldNames.safeMap { case (name, dtype) =>
+              val i = names.indexOf(name)
+              if (i == -1) {
+                GpuColumnVector.columnVectorFromNull(numRows, dtype)
+              } else {
+                val col = rawTable.getColumn(i)
+                doCast(col, getSparkType(col), dtype, false, false, false)
```

Comment:
schema should match this with the same number of columns and in the same order or we have a bug.  Use the types from struct instead of this. Or if we are going to turn it into a struct, make it into a struct before the cast, and then do the cast with just struct as the type.

Reply:
The value `dtype` is the type from `struct` schema. Where do we want to use the type from `struct` in `doCast`? If `dtype` is a struct, do we want to first cast `col` to `DType.STRUCT` then cast it to `dtype`? `doCast` only takes in Spark `DataType`, I will need to find the equivalent Spark type for the input cudf column `col` in some way.

Reply:
Oh, sorry I got confused. You are trying to get the from type, not the to type. I think what you have is fine then. It would be good to document it a bit to say that getSparkType is only used for the from type of a cast .

Reply:
Thank you! I added comments to document `getSparkType`.